### ISSUE TITLE
Idea: Syntactic sugar for checking whether a variable is one of multiple values

### ISSUE THREAD

**Issue by stevehjohn** [dotnet/csharplang#2155]
Hi,

**Problem**

I've always thought the mechanism for comparing a variable to multiple values to be a bit onerous/ugly/contains unnecessary repetition. For example:

```C#
if (response.HttpStatusCode == HttpStatusCode.BadRequest
    || response.HttpStatusCode == HttpStatusCode.Forbidden
    || response.HttpStatusCode == HttpStatusCode.InternalServerError)
{
    // Something went wrong - make it right (or at least log it!)
}
```

Of course, you could write an extension method to make it a bit prettier:

```C#
public static class GenericExtensions
{
    public static bool IsOneOf<T>(this T value, params T[] possibilities)
    {
        return possibilities.Contains(value);
    }
}
```

Then do:

```C#
if (response.HttpStatusCode.IsOneOf(HttpStatusCode.BadRequest, HttpStatusCode.Forbidden, HttpStatusCode.InternalServerError))
{
    // Something went wrong - make it right (or at least log it!)
}
```

**Proposal**

It might be nice to have some syntactic sugar built into the language for such operations. I'm not really sure exactly what it would look like. Maybe a `=[` and `!=[` operator? Perhaps something along the lines of:

```C#
if (response.HttpStatusCode =[ HttpStatusCode.BadRequest, HttpStatusCode.Forbidden, HttpStatusCode.InternalServerError ])
{
    // Something went wrong - make it right (or at least log it!)
}
```

And also the reverse, is not one of:

```C#
if (response.HttpStatusCode !=[ HttpStatusCode.OK, HttpStatusCode.Accepted, HttpStatusCode.NoContent])
{
    // Something went wrong - make it right (or at least log it!)
}
```

Regards,

Steve.

**Comment by Neme12** [comment]
With `or` patterns: #1350, you'd be able to write:
```c#
if (response.HttpStatusCode is HttpStatusCode.OK or HttpStatusCode.Accepted or HttpStatusCode.NoContent)
```
if you combined this with not patterns and parenthesized patterns, you'd have:
```c#
if (response.HttpStatusCode is not (HttpStatusCode.OK or HttpStatusCode.Accepted or HttpStatusCode.NoContent))
```

would you be happy with this syntax? if so, consider closing this in favor of #1350 and upvoting and/or commenting on that issue.

**Comment by YairHalberstadt** [comment]
Of also mention that switch expressions, which are almost definitely coming to C# 8 could help as well.

**Comment by ufcpp** [comment]
dup of https://github.com/dotnet/csharplang/issues/1787

**Comment by Neme12** [comment]
@YairHalberstadt How would switch expressions help? Can you show an example?

**Comment by Neme12** [comment]
> the is operator is currently in use for type checking.

Since C# 7.0, it's not only used for type checking. I suggest you read up on pattern matching. `and`, `or` and `not` patterns would introduce no changes to the `is` expression itself. It would just be another form of a pattern.

**Comment by YairHalberstadt** [comment]
@Neme12 
```csharp
if (response.HttpStatusCode switch 
            {
            HttpStatusCode.BadRequest => true,
            HttpStatusCode.Forbidden => true,
            HttpStatusCode.InternalServerError => true, 
            _ => false
            })
```

Is there an way to fall through a switch expression?

```csharp
if (response.HttpStatusCode switch 
            {
            HttpStatusCode.BadRequest,
            HttpStatusCode.Forbidden,
            HttpStatusCode.InternalServerError => true, 
            _ => false
            })
```
didn't work

**Comment by YairHalberstadt** [comment]
Closing as `or` patterns are implemented in C# 9.0, allowing you to do 

```csharp
if (response.HttpStatusCode is HttpStatusCode.BadRequest or response.HttpStatusCode or HttpStatusCode.Forbidden or HttpStatusCode.InternalServerError)
{
    // Something went wrong - make it right (or at least log it!)
}
```
